File: Leetcode/Algorithm/cpp/00336-Palindrome Pairs.cc
class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> ans;
        unordered_map<string, int> mp;
        int n = words.size();
        
        for (int i = 0; i < n; i++) {
            auto& word = words[i];
            mp[word] = i;
        }
        
        for (int i = 0; i < n; i++) {
            auto& word = words[i];
            vector<string> prefixs, suffixs;
            get_prefix(word, prefixs, suffixs);
            for (auto prefix: prefixs) {
                //printf("%s ", prefix.c_str());
                reverse(prefix.begin(), prefix.end());
                if (mp.find(prefix) != mp.end() && mp[prefix] != i) {
                    ans.push_back({i, mp[prefix]});
                }
            }
            for (auto suffix: suffixs) {
                reverse(suffix.begin(), suffix.end());
                if (mp.find(suffix) != mp.end() && mp[suffix] != i) {
                    ans.push_back({mp[suffix], i});
                }
            }
        }
        return ans;
    }
private:
    void get_prefix(const string& word, vector<string>& prefix, vector<string>& suffix) {
        memset(dp, 0, sizeof(dp));
        int n = word.size();
        for (int i = 0; i < n; i++) {
            dp[i][i] = 1;    
        }
        for (int i = 1; i < n; i++) {
            for (int j = 0; i + j < n; j++) {
                if (word[j] != word[i + j]) {
                    continue;
                }
                if (i > 1 && !dp[j + 1][i + j - 1]) {
                    continue;
                }
                dp[j][j + i] = 1;
            }
        }
        
        for (int i = 1; i < n; i++) {
            if (dp[i][n - 1]) {
                prefix.push_back(word.substr(0, i));
            }
        }
        prefix.push_back(word);
        
        for (int i = 0; i < n - 1; i++) {
            if (dp[0][i]) {
                suffix.push_back(word.substr(i + 1));
            }
        }
        
        if (dp[0][n - 1]) {
            prefix.push_back("");
            suffix.push_back("");
        }
    }
    char dp[222][222];
};

File: Leetcode/Algorithm/python/00336-Palindrome Pairs.py
MAGIC = 29
MOD = 0xdeadbeef

class Solution(object):
    def palindromePairs(self, words):
        d = dict([(self.get_hash(word), i) for i, word in enumerate(words)])
        res = []
        print d
        
        for i, word in enumerate(words):
            m = len(word)
            h1, h2, u = 0, 0, 1
            
            #print word
            for j in xrange(m + 1):
                #print h1, h2, word[j:], word[:m - j]
                if h1 in d and d[h1] != i and self.is_palindrome(word[j:]):  
                    res.append((i, d[h1]))
                    
                if h2 in d and d[h2] != i and self.is_palindrome(word[:m - j]):
                    res.append((d[h2], i))
                
                if j < m:
                    c1, c2 = word[j], word[m - j - 1]
                    h1 = ((u * (ord(c1) - ord('a') + 1)) + h1) % MOD
                    h2 = (h2 * MAGIC + ord(c2) - ord('a') + 1) % MOD
                    u = (u * MAGIC) % MOD
                    
        return list(set(res))

    def is_palindrome(self, word):
        i, j = 0, len(word) - 1
        while i <= j:
            if word[i] == word[j]:
                i += 1
                j -= 1
            else:
                return False
        return True 
        
    def get_hash(self, word):
        h1 = 0
        for c in word:
            h1 = (h1 * MAGIC + ord(c) - ord('a') + 1) % MOD
        return h1
            
        
